### PR TITLE
chore: added initial version of PPOTrainer support

### LINKED ISSUE #3522: [FEATURE] add `PPOTrainer` to `trl` integration `ArgillaTrainer`
**Is your feature request related to a problem? Please describe.**
We missed the `PPOTrainer` within our first integration leap and it would be best to add this too. 
https://github.com/lvwerra/trl#ppotrainer

image

**Describe the solution you'd like**
```python
task = TrainingTask.for_reward_modelling(...)
trainer = ArgillaTrainer(
   dataset=fds_dataset,
   task=task,
   framework="trl",
)
trainer.train()
trainer.save("reward_model")

# And then you can use this "reward_model" with PPO
task = TrainingTask.for_proximal_policy_optimization(...)
trainer = ArgillaTrainer(
   dataset=fds_dataset,
   task=task,
   framework="trl",
)
trainer.train(model=reward_model, generation_args)
```

**Describe alternatives you've considered**
N.A.

**Additional context**
N.A.

### PR DESCRIPTION
**Pull request by davidberenstein1957**
# Description

I added support for the PPOTrainer.

```python
from transformers import pipeline
from trl import PPOConfig

task_mapping = TrainingTask.for_proximal_policy_optimization(text=dataset.field_by_name("text"))
trainer = ArgillaTrainer(
    dataset=dataset,
    task=task_mapping,
    framework="trl",
    fetch_records=False
)
# assuming we have an arbitrarily trained textcat model
reward_model = pipeline("sentiment-analysis", model="lvwerra/distilbert-imdb")
trainer.update_config(reward_model=sentiment_pipe) # this is always required but if not done it provides a warning
trainer.train(output_dir="my_awesone_model")
```

Closes #3522 

**Type of change**

- [X] New feature (non-breaking change which adds functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [ ] `tests/integration/client/feedback/training/test_trl.py`

**Checklist**

- [ ] I added relevant documentation
- [X] I followed the style guidelines of this project
- [X] I did a self-review of my code
- [X] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/client/feedback/training/frameworks/trl.py [hunk 1]
```diff
@@ -218,12 +319,43 @@ def preprocess_function(examples) -> Dict[str, List]:
...
+
+                #### Compute rewards scores
+                texts = [q + r for q, r in zip(batch["query"], batch["response"])]
+                pipe_outputs = reward_model(texts, return_all_scores=True)
+                rewards = [torch.tensor(output[-1]["score"] * len(output)) for output in pipe_outputs]
+
+                #### Run PPO step
+                stats = self._trainer.step(query_tensors, response_tensors, rewards)
+                self._trainer.log_stats(stats, batch, rewards)
         else:
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/training/frameworks/trl.py:347]:
```suggestion
        else:
            self._metrics = None
```

File: src/argilla/client/feedback/training/schemas.py [hunk 2]
```diff
@@ -266,7 +280,7 @@ def for_direct_preference_optimization(
         ],
     ) -> "TrainingTaskForDPO":
         """
-        Return a task that can be used in `FeedbackDataset.prepare_for_training(framework="...", task)`
+        Return a task that can be used in `FeedbackDataset.prepare_for_training(prompt_chosen_rejected_func: Callable)`
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/training/schemas.py]:
```suggestion
        Provide `TrainingTask.for_direct_preference_optimization(prompt_chosen_rejected_func: Callable)
        Return a task that can be used in `FeedbackDataset.prepare_for_training(framework="...", task)`
```

File: src/argilla/client/feedback/training/schemas.py [hunk 3]
```diff
@@ -752,6 +814,10 @@ def for_reward_modelling(cls, *args, **kwargs) -> TrainingTaskForRM:
         cls.warn()
         return super().for_reward_modelling(*args, **kwargs)
 
+    @classmethod
+    def for_proximal_policy_optimization(cls, *args, **kwargs) -> TrainingTaskForPPO:
```
Comment (davidberenstein1957) [inline src/argilla/client/feedback/training/schemas.py:824]:
```suggestion
    def for_proximal_policy_optimization(cls, *args, **kwargs) -> TrainingTaskForPPO:
        cls.warn()
```

File: src/argilla/client/feedback/training/frameworks/trl.py [hunk 4]
```diff
@@ -204,7 +286,29 @@ def preprocess_function(examples) -> Dict[str, List]:
...
                 **self.trainer_kwargs,
             )
+        elif isinstance(self._task, TrainingTaskForPPO):
+            from datasets import concatenate_datasets
+
+            dataset = concatenate_datasets([x for x in [self._train_dataset, self._eval_dataset] if x is not None])
+
+            def tokenize(sample):
+                sample["input_ids"] = self._transformers_tokenizer.encode(sample["text"])
+                sample["query"] = self._transformers_tokenizer.decode(sample["input_ids"])
```
Comment (tomaarsen) [inline src/argilla/client/feedback/training/frameworks/trl.py]: Is there a difference between this and just `sample["text"]`?

Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/frameworks/trl.py]: Nope, this was some leftover form some code I copied from TRL and forgot to remove it.

File: src/argilla/client/feedback/training/frameworks/trl.py (hunk 1 above)
Comment (tomaarsen) [reply src/argilla/client/feedback/training/frameworks/trl.py:347]: Why should we reintroduce this variable? It's already defined a bit lower right?

File: src/argilla/client/feedback/training/frameworks/trl.py [hunk 5]
```diff
@@ -78,34 +118,57 @@ def __init__(
...
         }
         self.trainer_cls = self.trainer_mapping[type(self._task)]
 
         self.init_training_args()
 
+        if isinstance(self._task, TrainingTaskForPPO):
+            self._logger.warning(
+                "The PPOTrainer must be initialized by passing `reward_model`, `length_sampler_kwargs`, `generation_kwargs` as kwargs to the `update_config()`-method."
+            )
+
```
Comment (tomaarsen) [inline src/argilla/client/feedback/training/frameworks/trl.py]: Is this not duplicate? We also send this warning in `init_training_args` on the previous line

Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/frameworks/trl.py]: Yes, I think we want a logger to warn during `update_arguments` and an explicit raised exception during `train()`

File: src/argilla/client/feedback/training/frameworks/trl.py (hunk 1 above)
Comment (davidberenstein1957) [reply src/argilla/client/feedback/training/frameworks/trl.py:347]: Yes exactly, sorry I though I had removed this by accident.

File: src/argilla/client/feedback/training/frameworks/trl.py (hunk 5 above)
Comment (tomaarsen) [reply src/argilla/client/feedback/training/frameworks/trl.py]: Sounds good, I'll add that

**Comment by tomaarsen** [comment]
Test failures also exist in `feat/integration_trl` & fixed in develop. I'll propagate the fix down all branches, i.e. merge develop into `feat/integration_trl` and then update this branch as well.

**Comment by tomaarsen** [comment]
- [ ] `docs/_source/guides/llms/practical_guides/fine_tune.md`
- [x] `docs/_source/_common/tabs/train_update_config.md`

**Comment by tomaarsen** [comment]
Merged before documentation is complete, to allow @davidberenstein1957 to extend PPO further in `feat/integration_trl`.

**Pull request merged**